File: Kotlin/main/chap27/Section_4.kt
package chap27

import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import mocks.*

@Suppress("UNUSED_VARIABLE")
suspend fun demo(request: Request) =
  coroutineScope {
    val futureAd: Deferred<Ad> = async { fetchAd(request) }
    val futureData: Deferred<Data> = async { dbLookup(request) }
    val data: Data = futureData.await()
    val ad: Ad = futureAd.await()
    val page: Page = makePage(data, ad)
  }

File: Scala/main/chap27/Section_4.scala
package chap27

import cps.compat.sip22.{ async, await }
import mocks.Chap25.*

import scala.concurrent.{ ExecutionContext, Future }

//noinspection DuplicatedCode
object Section_4:
   def demo(request: Request)(using ExecutionContext): Future[Page] = async {
      val futureAd: Future[Ad]     = Future(fetchAd(request))
      val futureData: Future[Data] = Future(dbLookup(request))
      val ad: Ad                   = await(futureAd)
      val data: Data               = await(futureData)
      val page: Page               = makePage(data, ad)

      page
   }
